From f907f1113cbb7c97c8c4ed251731f69c6c140e56 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:24:58 -0400
Subject: [tools] perf: Save failed result of get_srcline()

Message-id: <1408462094-14194-130-git-send-email-jolsa@redhat.com>
Patchwork-id: 88107
O-Subject: [PATCH RHEL7.1 BZ1131394 129/325] perf tools: Save failed result of get_srcline()
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 2cc9d0ef577975abb3ebce7d5978559ec1c73633
Author: Namhyung Kim <namhyung.kim@lge.com>
Date: Wed Sep 11 14:09:31 2013 +0900

description
===========
Some dso's lack srcline info, so there's no point to keep trying on
them.  Just save failture status and skip them.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/dso.c b/tools/perf/util/dso.c
index 6bfc8aa..af4c687 100644
--- a/tools/perf/util/dso.c
+++ b/tools/perf/util/dso.c
@@ -450,6 +450,7 @@ struct dso *dso__new(const char *name)
   dso->rel = 0;
   dso->sorted_by_name = 0;
   dso->has_build_id = 0;
+  dso->has_srcline = 1;
   dso->kernel = DSO_TYPE_USER;
   dso->needs_swap = DSO_SWAP__UNSET;
   INIT_LIST_HEAD(&dso->node);
diff --git a/tools/perf/util/dso.h b/tools/perf/util/dso.h
index 72eedd6..9ac666a 100644
--- a/tools/perf/util/dso.h
+++ b/tools/perf/util/dso.h
@@ -83,6 +83,7 @@ struct dso {
  enum dso_binary_type data_type;
  u8   adjust_symbols:1;
  u8   has_build_id:1;
+ u8   has_srcline:1;
  u8   hit:1;
  u8   annotate_warned:1;
  u8   sname_alloc:1;
diff --git a/tools/perf/util/srcline.c b/tools/perf/util/srcline.c
index c736d94..dcff10b 100644
--- a/tools/perf/util/srcline.c
+++ b/tools/perf/util/srcline.c
@@ -58,10 +58,13 @@ char *get_srcline(struct dso *dso, unsigned long addr)
 {
  char *file;
  unsigned line;
- char *srcline = SRCLINE_UNKNOWN;
+ char *srcline;
  char *dso_name = dso->long_name;
  size_t size;
 
+ if (!dso->has_srcline)
+  return SRCLINE_UNKNOWN;
+
  if (dso_name[0] == '[')
   goto out;
 
@@ -81,8 +84,11 @@ char *get_srcline(struct dso *dso, unsigned long addr)
   srcline = SRCLINE_UNKNOWN;
 
  free(file);
-out:
  return srcline;
+
+out:
+ dso->has_srcline = 0;
+ return SRCLINE_UNKNOWN;
 }
 
 void free_srcline(char *srcline)
-- 
1.7.1